### ISSUE TITLE
`ConcurrentModificationException` while resolving build dependencies

### ISSUE THREAD

**Issue by gmazzo** [avast/gradle-docker-compose-plugin#352]
In our setup, we have some kind of hierarchy of conventions and if we combine `jvm-test-suite` and a reactive `plugins.withId("com.avast.gradle.docker-compose")` we start getting this weird message while Gradle is trying to resolve a task dependencies:
```sh
Could not determine the dependencies of task ':payments-admin:compileJava'.
> Could not resolve all dependencies for configuration ':payments-admin:compileClasspath'.
   > java.util.ConcurrentModificationException (no error message)
```

The relevant piece of code is this:
```kotlin
    val intTest by registering(JvmTestSuite::class) {

        targets.configureEach {
            plugins.withId("com.avast.gradle.docker-compose") {
                the<ComposeExtension>()
                    .nested("it")
                    .isRequiredBy(testTask)
            }
        }
    }
```

After some research, we found out the root cause is the logic under `isRequiredBy` when a `TaskProvider`  is used , because it's doing this as `isRequiredByCore` (with `fromConfigure=true`):
```kotlin
        project.tasks.findAll { Task.class.isAssignableFrom(it.class) && ((Task) it).name.toLowerCase().contains('classes') }
                .each { classesTask ->
                    if (fromConfigure) {
                        upTask.get().shouldRunAfter classesTask
                    } else {
                        upTask.configure { it.shouldRunAfter classesTask }
                    }
                }
```

Can you state the purpose of this logic? I see two problems in here:
1) Looking for tasks by class and name is error-prone to me. If there is some kind of dependencies between the module's classes and the `composeUp` task, it should be explicit.
2) Did you know that `project.tasks.findAll` is actually a [build performance killer](https://docs.gradle.org/4.9/userguide/task_configuration_avoidance.html#sec:old_vs_new_configuration_api_overview)? It's realizing all tasks in the project when you do that, so having used `registered` and `configureEach` is effectible no-op.

I think the issue is that we are changing the tasks's build dependencies in an unsafe way, indirectly by eagerly realizing all tasks. This issue is gone if you either:
1) Use `isRequiredBy(testTask.get())` which is not correct to me, because we are creating the `Test` task at configuration time.
2) Replace `isRequiredBy` with our own logic, removing the `project.tasks.findAll` part:
```kotlin
    val intTest by registering(JvmTestSuite::class) {
        targets.configureEach {
            plugins.withId("com.avast.gradle.docker-compose") {
                val itCompose = the<ComposeExtension>().nested("it")

                testTask.configure test@{
                    dependsOn(itCompose.upTask)
                    finalizedBy(itCompose.downTask)
                    itCompose.exposeAsSystemProperties(this@test)
                }
            }
        }
    }
```

**Comment by augi** [comment]
Hello,

the reason for this code is that if you iterate and you have a bug in your source code syntax (so `classes` task would fail) then you have to wait for `composeUp`. So by suggesting Gradle to execute `composeUp` after `classes`, you don't have to wait for a long time for a compilation error.

Feel free to prepare a PR that would remove this optimization, if it harms you 🙏

**Comment by gmazzo** [comment]
> Feel free to prepare a PR that would remove this optimization, if it harms you 🙏
What about this https://github.com/avast/gradle-docker-compose-plugin/pull/356?

**Comment by augi** [comment]
Fixed in `0.16.4`.
